wgplugin.cpp

```diff
@@ -0,0 +1,231 @@
+ * if advised of the possibility of such damage.
+ *
+ */
+
+#include <iostream>
+#include <cstring>
+
+#include "wgplugin.h"
+#include "flowifc.h"
+#include "flowcacheplugin.h"
+#include "packet.h"
+#include "ipfixprobe.h"
+#include "ipfix-elements.h"
+
+using namespace std;
+
+#define WG_UNIREC_TEMPLATE "WG_CONF_LEVEL,WG_SRC_PEER,WG_DST_PEER"
+
+UR_FIELDS (
+   uint8 WG_CONF_LEVEL,
+   uint32 WG_SRC_PEER,
+   uint32 WG_DST_PEER
+)
+
+WGPlugin::WGPlugin(const options_t &module_options)
+{
+   flow_flush = false;
+   print_stats = module_options.print_stats;
+   total = 0;
+   identified = 0;
+}
+
+WGPlugin::WGPlugin(const options_t &module_options, vector<plugin_opt> plugin_options) : FlowCachePlugin(plugin_options)
+{
+   flow_flush = false;
+   print_stats = module_options.print_stats;
+   total = 0;
+   identified = 0;
+}
+
+int WGPlugin::post_create(Flow &rec, const Packet &pkt)
+{
+   if (pkt.ip_proto == IPPROTO_UDP) {
+      add_ext_wg(pkt.payload, pkt.payload_length, pkt.source_pkt, rec);
+   }
+
+   return 0;
+}
+
+int WGPlugin::pre_update(Flow &rec, Packet &pkt)
+{
+   RecordExtWG *vpn_data = (RecordExtWG *) rec.getExtension(wg);
+   if (vpn_data != NULL) {
+      parse_wg(pkt.payload, pkt.payload_length, pkt.source_pkt, vpn_data);
+
+      if (flow_flush) {
+         flow_flush = false;
+         return FLOW_FLUSH_WITH_REINSERT;
+      }
+   }
+
+   return 0;
+}
+
+
+void WGPlugin::pre_export(Flow &rec)
+{
+}
+
+void WGPlugin::finish()
+{
+   if (print_stats) {
+      cout << "WG plugin stats:" << endl;
+      cout << "   Identified WG packets: " << identified << endl;
+      cout << "   Total packets processed: " << total << endl;
+   }
+}
+
+const char *ipfix_wg_template[] = {
+   IPFIX_WG_TEMPLATE(IPFIX_FIELD_NAMES)
+   NULL
+};
+
+const char **WGPlugin::get_ipfix_string()
+{
+   return ipfix_wg_template;
+}
+
+string WGPlugin::get_unirec_field_string()
+{
+   return WG_UNIREC_TEMPLATE;
+}
+
+bool WGPlugin::include_basic_flow_fields()
+{
+   return true;
+}
+
+bool WGPlugin::parse_wg(const char *data, unsigned int payload_len, bool source_pkt, RecordExtWG *ext)
+{
+   uint32_t cmp_peer;
+   uint32_t cmp_new_peer;
+
+   total++;
+
+   // The smallest message (according to specs) is the data message (0x04) with 16 header bytes
+   // and 16 bytes of (empty) data authentication.
+   // Anything below that is not a valid WireGuard message.
+   if (payload_len < WG_PACKETLEN_MIN_TRANSPORT_DATA) {
+      return false;
+   }
+
+   // Let's try to parse according to the first 4 bytes, and see if that is enough.
+   // The first byte is 0x01-0x04, the following three bytes are reserved (0x00).
+   uint8_t pkt_type = data[0];
+   if (pkt_type < WG_PACKETTYPE_INIT_TO_RESP || pkt_type > WG_PACKETTYPE_TRANSPORT_DATA) {
+      return false;
+   }
+   if (data[1] != 0x0 || data[2] != 0x0 || data[3] != 0x0) {
+      return false;
+   }
+
+   // Next, check the packet contents based on the message type.
+   switch(pkt_type) {
```

Comment:
Fix coding style in this method.
`switch(` -> `switch (`
`(data+4)` -> `(data + 4)`
`(! source_pkt)` -> `(!source_pkt)`